### DISCUSSION TITLE
Design of system with many tables linked together

### DISCUSSION THREAD

**Discussion by beolafsen** [oqtane/oqtane.framework#2901] [Q&A]
I have a design question that I've been wondering about for a while. My system consists of many tables that are linked together. As of now approx. 20-30 tables, but constantly growing.
To maintain the data in the tables, I have created a Module per table. To link the data in the tables together, I use SiteId (not ModuleId).
I think the best solution would have been to use ModuleId to link the tables together, but then all the UIs would have to be in the same module - something I'm not sure if it can be solved in practice.

How is this thought to be solved in Oqtane?

**Comment by sbwalker** [comment]
The "scoping" of data really depends on the requirements of your specific application. Oqtane provides a variety of scoping mechanisms:

No Scope - if you do not intend to use the multi-tenancy aspect of Oqtane then you do not need to scope your data at all - which would mean that it is "global" to the application

Site Scope - if you want to use multi-tenancy then you should scope your data to a SiteId (ie. ModuleState.SiteId). This will allow you to select the data which is related to a specific site in your installation. This level of scoping is appropriate if only a single instance of a module is expected to be rendered in a site. Most of the Admin modules in Oqtane follow this approach ie. the Role Management module is only expected to exist on the Role Management page in a site - not on any other pages. 

Module Scope - there may be scenarios where you wish to render multiple instances of the same module on the same page or on multiple pages in your site - in this case you would want to scope your data based on ModuleId (ie. ModuleState.ModuleId). This will allow you to select the data which is related to a specific module instance on a page in a site. An example of this type of module is the Html/Text module which is expected to have many module instances. 

Note there is no correlation between how you scope your data and how you organize your module component files. For example, the Oqtane Client project is essentially a large Oqtane module project. It is organized into folders/namespaces which define the functionality. It contains individual modules which are scoped at both the Site and Module level depending on how they are used. By bundling all of these modules into a single project also means that it is also deployed as a single unit - which may or may not be your preference.
